drivers/misc/mcux_flexio/mcux_flexio.c

```diff
@@ -0,0 +1,253 @@
+/*
+ * Copyright (c) 2024, STRIM, ALC
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT nxp_flexio
+
+#include <zephyr/kernel.h>
+#include <zephyr/device.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/drivers/clock_control.h>
+#include <fsl_flexio.h>
+
+LOG_MODULE_REGISTER(mcux_flexio, CONFIG_MCUX_FLEXIO_LOG_LEVEL);
+
+#include "mcux_flexio.h"
+
+
+struct mcux_flexio_config {
+	FLEXIO_Type *base;
+	const struct device *clock_dev;
+	clock_control_subsys_t clock_subsys;
+	void (*irq_config_func)(const struct device *dev);
+	void (*irq_enable_func)(void);
+	void (*irq_disable_func)(void);
+};
+
+typedef const struct mcux_flexio_child *mcux_flexio_map_child_t;
+
+struct mcux_flexio_data {
+	struct k_mutex lock;
+	uint32_t shifter_indexes_used;
+	uint32_t timer_indexes_used;
+	mcux_flexio_map_child_t *map_shifter_child;
+	mcux_flexio_map_child_t *map_timer_child;
+	uint32_t map_shifter_child_count;
+	uint32_t map_timer_child_count;
+};
+
+
+static int mcux_flexio_child_take_shifter_idx(const struct device *dev)
+{
+	struct mcux_flexio_data *data = dev->data;
+
+	for (uint32_t i = 0; i < data->map_shifter_child_count; i++) {
+		if ((data->shifter_indexes_used & BIT(i)) == 0) {
+			WRITE_BIT(data->shifter_indexes_used, i, 1);
+			return i;
+		}
+	}
+
+	return -ENOBUFS;
+}
+
+static int mcux_flexio_child_take_timer_idx(const struct device *dev)
+{
+	struct mcux_flexio_data *data = dev->data;
+
+	for (uint32_t i = 0; i < data->map_timer_child_count; i++) {
+		if ((data->timer_indexes_used & BIT(i)) == 0) {
+			WRITE_BIT(data->timer_indexes_used, i, 1);
+			return i;
+		}
+	}
+
+	return -ENOBUFS;
+}
+
+static void mcux_flexio_isr(const struct device *dev)
+{
+	const struct mcux_flexio_config *config = dev->config;
+	struct mcux_flexio_data *data = dev->data;
+	FLEXIO_Type *base = config->base;
+
+	mcux_flexio_map_child_t *map_shifter_child = data->map_shifter_child;
+	uint32_t map_shifter_child_count = data->map_shifter_child_count;
+	uint32_t shifter_status_flag = FLEXIO_GetShifterStatusFlags(base);
+	uint32_t shifter_error_flag = FLEXIO_GetShifterErrorFlags(base);
+
+	if (shifter_status_flag || shifter_error_flag) {
+		for (uint32_t idx = 0; idx < map_shifter_child_count; idx++) {
+			if (((shifter_status_flag | shifter_error_flag) & BIT(idx)) != 0) {
+				const struct mcux_flexio_child *child = map_shifter_child[idx];
+
+				if (child != NULL) {
+					child->isr(child->user_data);
+				}
```

Comment:
We should check if the isr field is not NULL before calling it. Not every driver might be interested in the ISR as they may use DMA.